### ISSUE TITLE
How to reproduce 72.2 IoU on validation set?

### ISSUE THREAD

**Issue by xternalz** [Eromera/erfnet_pytorch#1]
Hi, I tried training ERFNet with the default configuration (single GPU):

> python main.py --savedir erfnet_training1 --datadir /home/datasets/cityscapes/ --num-epochs 150 --batch-size 6

But, the best validation mean IoU that I could get was just ~69.0% (instead of 72.2 achieved by your pretrained model). Could you point out what could be causing this disparity?

*Another quick question - Why is the evaluation done on images and labels both resized to height of 512 pixels, rather than original resolution of height of 1024 pixels?

Thanks!

**Comment by Eromera** [comment]
Hi!

That default configuration that you trained is "from scratch", this is: training the encoder on downsampled 1/8 cityscapes labels, then attach decoder and train again the full network. 69-70% is a reasonable result for the "from scratch" training, since the weights from scratch generalize a bit worse in val/test sets.

The 72.2% IoU is achieved by training the network using an encoder pretrained in ImageNet. I provided this "pretrained encoder" in the trained_models folder so you only need to add the "--decoder" flag and the "--pretrainedEncoder" argument pointing to that encoder. You can just use it like this:

python main.py --savedir erfnet_training1 --datadir /home/datasets/cityscapes/ --num-epochs 150 --batch-size 6 --decoder --pretrainedEncoder "../trained_models/erfnet_encoder_pretrained.pth.tar"

About your other question, if the model is trained using 1024x512 (default) then you should also evaluate on 1024x512 for best accuracy. The training code shows IoU in the resolution used for training. The models trained in the paper are trained in 1024x512 and then the predictions are rescaled to 2048x1024 by simple interpolation to evaluate in the cityscapes server, since all models are evaluated at that resolution. You could also train the model at 2048x1024 and evaluate it in that same resolution (without rescaling anything), but it would be slower and in my opinion there is not a big gain in accuracy compared to all the efficiency that you lose (almost 4x slower), so it's actually a trade-off. You can try the eval scripts with any size you prefer, you can use the "eval_cityscapes_server.py" to create the predictions for the val/test set and upload them to the cityscapes server (the input will be rescaled to the ratio that you specify and then the output will be rescaled back to 2048x1024), or you can use the new "eval_iou.py" code to check the iou directly in the val set.

Thanks for messaging!

**Comment by xternalz** [comment]
Thank you for providing such awesome clarifications!
